File: transfers/legit/blueprints/plans.py
from dataclasses import dataclass
from datetime import datetime

from common.timeline import active_months
import entities.models as models
import entities.personas as personas_mod

from .models import Overrides, Timeline, Network, Macro


@dataclass(frozen=True, slots=True)
class CounterpartyPlan:
    hub_accounts: list[str]
    hub_set: frozenset[str]
    employers: list[str]
    landlords: list[str]
    biller_accounts: list[str]
    issuer_acct: str


@dataclass(frozen=True, slots=True)
class PersonaPlan:
    persona_for_person: dict[str, str]
    persona_objects: dict[str, models.Persona]
    persona_names: list[str]


@dataclass(frozen=True, slots=True)
class LegitBuildPlan:
    start_date: datetime
    days: int
    seed: int
    all_accounts: list[str]
    persons: list[str]
    month_starts: list[datetime]
    primary_acct_for_person: dict[str, str]
    counterparties: CounterpartyPlan
    personas: PersonaPlan

    @property
    def paydays(self) -> list[datetime]:
        """
        Compatibility shim for older generators that still expect `plan.paydays`.
        These are month anchors, not true payroll schedules.
        """
        return self.month_starts


def _select_hub_accounts(
    timeline: Timeline,
    network: Network,
    macro: Macro,
) -> list[str]:
    persons = list(network.accounts.by_person)
    if not persons:
        return []

    n_hubs = int(macro.pop.size * macro.hubs.fraction)
    n_hubs = max(1, min(n_hubs, len(persons)))

    hub_people = timeline.rng.choice_k(persons, n_hubs, replace=False)
    return [network.accounts.by_person[person_id][0] for person_id in hub_people]


def _primary_acct_for_person(
    accounts: models.Accounts,
) -> dict[str, str]:
    return {
        person_id: acct_ids[0]
        for person_id, acct_ids in accounts.by_person.items()
        if acct_ids
    }


def _build_counterparty_plan(
    timeline: Timeline,
    network: Network,
    macro: Macro,
    overrides: Overrides,
) -> CounterpartyPlan:
    all_accounts = network.accounts.ids
    if not all_accounts:
        raise ValueError("network.accounts.ids must be non-empty")

    hub_accounts = _select_hub_accounts(timeline, network, macro)
    hub_set = frozenset(hub_accounts)

    fallback_acct = hub_accounts[0] if hub_accounts else all_accounts[0]

    pools = overrides.counterparty_pools

    if pools is not None and pools.employer_ids:
        employers = list(pools.employer_ids)
    else:
        employers = (
            hub_accounts[: max(1, len(hub_accounts) // 5)]
            if hub_accounts
            else [fallback_acct]
        )

    if pools is not None and pools.landlord_ids:
        landlords = list(pools.landlord_ids)
    else:
        landlords = hub_accounts if hub_accounts else [fallback_acct]

    biller_accounts = hub_accounts if hub_accounts else [fallback_acct]
    issuer_acct = fallback_acct

    return CounterpartyPlan(
        hub_accounts=hub_accounts,
        hub_set=hub_set,
        employers=employers,
        landlords=landlords,
        biller_accounts=biller_accounts,
        issuer_acct=issuer_acct,
    )


def _build_persona_plan(
    timeline: Timeline,
    macro: Macro,
    overrides: Overrides,
    persons: list[str],
) -> PersonaPlan:
    persona_for_person = overrides.persona_for_person
    if persona_for_person is None:
        persona_for_person = personas_mod.assign(
            macro.personas,
            timeline.rng,
            persons,
        )

    persona_objects = overrides.persona_objects
    if persona_objects is None:
        persona_objects = {
            pid: personas_mod.get_persona(name)
            for pid, name in persona_for_person.items()
        }

    persona_names = overrides.persona_names
    if persona_names is None:
        persona_names = list(personas_mod.PERSONAS.keys())

    return PersonaPlan(
        persona_for_person=persona_for_person,
        persona_objects=persona_objects,
        persona_names=persona_names,
    )


def build_legit_plan(
    timeline: Timeline,
    network: Network,
    macro: Macro,
    overrides: Overrides,
) -> LegitBuildPlan:
    start_date = timeline.window.start_date
    days = int(timeline.window.days)
    seed = int(macro.pop.seed)
    persons = list(network.accounts.by_person)

    counterparties = _build_counterparty_plan(timeline, network, macro, overrides)
    personas = _build_persona_plan(timeline, macro, overrides, persons)
    primary_acct_for_person = _primary_acct_for_person(network.accounts)
    month_starts = active_months(start_date, days)

    return LegitBuildPlan(
        start_date=start_date,
        days=days,
        seed=seed,
        all_accounts=network.accounts.ids,
        persons=persons,
        month_starts=month_starts,
        primary_acct_for_person=primary_acct_for_person,
        counterparties=counterparties,
        personas=personas,
    )

File: src/transfers/legit/blueprints/plans.cpp
#include "phantomledger/transfers/legit/blueprints/plans.hpp"

#include "phantomledger/entities/landlords.hpp"
#include "phantomledger/entities/synth/personas/make.hpp"
#include "phantomledger/primitives/time/calendar.hpp"
#include "phantomledger/primitives/time/window.hpp"
#include "phantomledger/taxonomies/personas/names.hpp"

#include <algorithm>
#include <cstddef>
#include <cstdint>
#include <stdexcept>
#include <unordered_map>
#include <utility>
#include <vector>

namespace PhantomLedger::transfers::legit::blueprints {

namespace {

[[nodiscard]] std::size_t hubCountFor(const PopulationTuning &population,
                                      std::size_t personCount) noexcept {
  if (personCount == 0) {
    return 0;
  }

  const auto populationCount = population.count == 0
                                   ? personCount
                                   : static_cast<std::size_t>(population.count);

  const double fraction = std::clamp(population.hubFraction, 0.0, 0.5);

  const auto requested =
      static_cast<std::size_t>(static_cast<double>(populationCount) * fraction);

  return std::clamp(requested, std::size_t{1}, personCount);
}

[[nodiscard]] std::vector<entity::Key>
selectHubAccounts(const PlanRequest &request,
                  const std::vector<entity::PersonId> &persons) {
  const auto &census = request.census;

  if (persons.empty() || request.rng == nullptr || census.accounts == nullptr ||
      census.ownership == nullptr) {
    return {};
  }

  const auto count = hubCountFor(request.population, persons.size());
  if (count == 0) {
    return {};
  }

  const auto chosenIdx =
      request.rng->choiceIndices(persons.size(), count, /*replace=*/false);

  std::vector<entity::Key> out;
  out.reserve(chosenIdx.size());

  for (const auto idx : chosenIdx) {
    const auto person = persons[idx];
    const auto recordIx = census.ownership->primaryIndex(person);

    if (recordIx < census.accounts->records.size()) {
      out.push_back(census.accounts->records[recordIx].id);
    }
  }

  return out;
}

[[nodiscard]] std::unordered_map<entity::PersonId, std::uint32_t>
primaryAcctRecordIxByPerson(const CensusSource &census) {
  std::unordered_map<entity::PersonId, std::uint32_t> out;

  if (census.accounts == nullptr || census.ownership == nullptr) {
    return out;
  }

  const auto &ownership = *census.ownership;
  const std::size_t personCount = ownership.byPersonOffset.empty()
                                      ? 0
                                      : ownership.byPersonOffset.size() - 1;

  out.reserve(personCount);

  for (entity::PersonId person = 1;
       person <= static_cast<entity::PersonId>(personCount); ++person) {
    const auto start = ownership.byPersonOffset[person - 1];
    const auto end = ownership.byPersonOffset[person];

    if (start == end) {
      continue;
    }

    out.emplace(person, ownership.primaryIndex(person));
  }

  return out;
}

struct LandlordResolution {
  std::vector<entity::Key> ids;
  std::unordered_map<entity::Key, entity::landlord::Type> typeOf;
};

[[nodiscard]] LandlordResolution
resolveLandlords(const CounterpartySource &source,
                 const std::vector<entity::Key> &hubAccounts,
                 entity::Key fallbackAcct) {
  LandlordResolution out;

  if (source.landlords != nullptr && !source.landlords->records.empty()) {
    out.ids.reserve(source.landlords->records.size());
    out.typeOf.reserve(source.landlords->records.size());

    for (const auto &record : source.landlords->records) {
      out.ids.push_back(record.accountId);
      out.typeOf.emplace(record.accountId, record.type);
    }

    return out;
  }

  if (!hubAccounts.empty()) {
    out.ids = hubAccounts;
  } else {
    out.ids.push_back(fallbackAcct);
  }

  return out;
}

[[nodiscard]] CounterpartyPlan
buildCounterpartyPlan(const PlanRequest &request,
                      const std::vector<entity::PersonId> &persons) {
  const auto *accounts = request.census.accounts;

  if (accounts == nullptr || accounts->records.empty()) {
    throw std::invalid_argument("PlanRequest.census.accounts must be non-empty "
                                "to build counterparties");
  }

  CounterpartyPlan plan;

  plan.hubAccounts = selectHubAccounts(request, persons);
  plan.hubSet.reserve(plan.hubAccounts.size());
  plan.hubSet.insert(plan.hubAccounts.begin(), plan.hubAccounts.end());

  const auto fallbackAcct = !plan.hubAccounts.empty()
                                ? plan.hubAccounts.front()
                                : accounts->records.front().id;

  const auto *counterparties = request.counterparties.directory;

  if (counterparties != nullptr &&
      !counterparties->employers.accounts.all.empty()) {
    plan.employers = counterparties->employers.accounts.all;
  } else if (!plan.hubAccounts.empty()) {
    const auto take = std::max<std::size_t>(1, plan.hubAccounts.size() / 5);
    plan.employers.assign(plan.hubAccounts.begin(),
                          plan.hubAccounts.begin() + take);
  } else {
    plan.employers.push_back(fallbackAcct);
  }

  auto landlords =
      resolveLandlords(request.counterparties, plan.hubAccounts, fallbackAcct);
  plan.landlords = std::move(landlords.ids);
  plan.landlordTypeOf = std::move(landlords.typeOf);

  plan.billerAccounts = !plan.hubAccounts.empty()
                            ? plan.hubAccounts
                            : std::vector<entity::Key>{fallbackAcct};

  plan.issuerAcct = fallbackAcct;

  return plan;
}

[[nodiscard]] PersonaPlan
buildPersonaPlan(const PlanRequest &request,
                 const std::vector<entity::PersonId> &persons) {
  PersonaPlan plan;

  if (request.personas.pack != nullptr) {
    plan.pack = request.personas.pack;
  } else {
    if (request.rng == nullptr) {
      throw std::invalid_argument("buildLegitPlan requires either "
                                  "PersonaSource.pack or PlanRequest.rng");
    }

    const auto popSize = static_cast<std::uint32_t>(persons.size());
    const auto baseSeed = static_cast<std::uint64_t>(request.population.seed);

    plan.ownedPack =
        entities::synth::personas::makePack(*request.rng, popSize, baseSeed);
    plan.pack = &*plan.ownedPack;
  }

  plan.personaNames.reserve(::PhantomLedger::personas::kKindCount);

  for (std::size_t i = 0; i < ::PhantomLedger::personas::kKindCount; ++i) {
    const auto kind = static_cast<::PhantomLedger::personas::Type>(i);
    plan.personaNames.push_back(::PhantomLedger::personas::name(kind));
  }

  return plan;
}

[[nodiscard]] std::vector<entity::PersonId>
extractPersons(const CensusSource &census) {
  std::vector<entity::PersonId> out;

  if (census.accounts == nullptr || census.ownership == nullptr) {
    return out;
  }

  const auto &ownership = *census.ownership;
  if (ownership.byPersonOffset.size() <= 1) {
    return out;
  }

  const auto personCount = ownership.byPersonOffset.size() - 1;
  out.reserve(personCount);

  for (entity::PersonId person = 1;
       person <= static_cast<entity::PersonId>(personCount); ++person) {
    const auto start = ownership.byPersonOffset[person - 1];
    const auto end = ownership.byPersonOffset[person];

    if (start != end) {
      out.push_back(person);
    }
  }

  return out;
}

} // namespace

LegitBuildPlan buildLegitPlan(const PlanRequest &request) {
  LegitBuildPlan plan;

  plan.startDate = request.window.start;
  plan.days = static_cast<std::int32_t>(request.window.days);
  plan.seed = static_cast<std::uint64_t>(request.population.seed);

  plan.allAccounts = request.census.accounts;
  plan.persons = extractPersons(request.census);

  plan.counterparties = buildCounterpartyPlan(request, plan.persons);

  plan.personas = buildPersonaPlan(request, plan.persons);

  plan.primaryAcctRecordIx = primaryAcctRecordIxByPerson(request.census);

  const auto endExcl = time::addDays(request.window.start, plan.days);
  plan.monthStarts = time::monthStarts(request.window.start, endExcl);

  return plan;
}

} // namespace PhantomLedger::transfers::legit::blueprints
